### ISSUE TITLE
Multiple compilation errors

### ISSUE THREAD

**Issue by agvxov** [jeaye/color_coded#203]
**tl;dr:**
The project only compiles with C++17, but cmake uses C++14.

**Issue:**
`
$ git clone https://github.com/jeaye/color_coded.git
Cloning into 'color_coded'...
remote: Enumerating objects: 4859, done.
remote: Counting objects: 100% (177/177), done.
remote: Compressing objects: 100% (104/104), done.
remote: Total 4859 (delta 86), reused 132 (delta 47), pack-reused 4682
Receiving objects: 100% (4859/4859), 8.18 MiB | 20.93 MiB/s, done.
Resolving deltas: 100% (2474/2474), done.
$ cd color_coded/
$ mkdir build && cd build
$ cmake .. -DDOWNLOAD_CLANG=0   # downloading clang footguns the process during configuration with another type of error
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.

-- The C compiler identification is GNU 12.2.1
-- The CXX compiler identification is GNU 12.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found 64bit system
-- Updating submodules
Submodule 'lib/jest' (https://github.com/jeaye/jest.git) registered for path 'lib/jest'
Cloning into '/home/anon/Swap/color_coded/lib/jest'...
Submodule path 'lib/jest': checked out '3506361cc94b44406b997e0507a362864687e982'
-- Performing Test COMPILER_SUPPORTS_CXX14
-- Performing Test COMPILER_SUPPORTS_CXX14 - Success
-- Performing Test COMPILER_SUPPORTS_CXX1Y
-- Performing Test COMPILER_SUPPORTS_CXX1Y - Success
-- Looking for Lua
-- Found Lua
--   Lua include /usr/local/include
--   Lua lib /usr/lib64/liblua5.1.so;/usr/lib64/libm.so
-- Looking for cbreak in /usr/lib64/libncurses.so
-- Looking for cbreak in /usr/lib64/libncurses.so - not found
-- Looking for nodelay in /usr/lib64/libncurses.so
-- Looking for nodelay in /usr/lib64/libncurses.so - not found
-- Found Curses: /usr/lib64/libncurses.so
-- Found ZLIB: /usr/local/lib/libz.so (found version "1.2.13")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found LLVM: /usr/lib/llvm/16 (found version "16.0.0")
-- Detected that llvm-config comes from a build-tree, adding more include directories for Clang
-- Found Clang (LLVM version: 16.0.0)
--   Include dirs:       /usr/lib/llvm/16/include;/tools/clang/include;/tools/clang/include
--   Clang libraries:    /usr/lib/llvm/16/lib64/libclang.so
--   Libclang C library: /usr/lib/llvm/16/lib64/libclang.so
-- Generating sources
-- Checking for luajit
-- Configuring done (1.6s)
-- Generating done (0.0s)
-- Build files have been written to: /home/anon/Swap/color_coded/build
$ make &> make.log
`
[make.log](https://github.com/jeaye/color_coded/files/12044513/make.log)

Judging from all the errors surounding std::optional i suspected the compiler standard version was at fault, so i retried with a clean download and modified 'cmake/cxx14.cmake', blindly to the following:
`
include(CheckCXXCompilerFlag)
CHECK_CXX_COMPILER_FLAG("-std=c++17" COMPILER_SUPPORTS_CXX14)
CHECK_CXX_COMPILER_FLAG("-std=c++17" COMPILER_SUPPORTS_CXX1Y)
if(COMPILER_SUPPORTS_CXX14)
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
elseif(COMPILER_SUPPORTS_CXX1Y)
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
else()
  message(FATAL_ERROR "${CMAKE_CXX_COMPILER} has no C++14 support")
endif()
`

This makes the project compile, see:
[make2.log](https://github.com/jeaye/color_coded/files/12044571/make2.log)
leaving only a linking error which is presumably a poorly reproducible problem, specific to my machine, that i expect support with.

**Set up:**
[setup.txt](https://github.com/jeaye/color_coded/files/12044675/setup.txt)
